File: backend/src/application/use-cases/RegisterUser.ts
import { User, UserRole, VulnerabilityType } from '../../domain/entities/User';
import { IUserRepository } from '../../domain/repositories/IUserRepository';

export interface RegisterUserDTO {
  email: string;
  password: string;
  fullName: string;
  phone: string;
  skills?: string[];
  vulnerabilities?: VulnerabilityType[];
  resources?: string[];
  latitude?: number;
  longitude?: number;
}

export interface IPasswordHasher {
  hash(password: string): Promise<string>;
}

export class RegisterUser {
  constructor(
    private readonly userRepository: IUserRepository,
    private readonly passwordHasher: IPasswordHasher
  ) {}

  async execute(dto: RegisterUserDTO): Promise<User> {
    // Check if user already exists
    const existing = await this.userRepository.findByEmail(dto.email);
    if (existing) {
      throw new Error('User with this email already exists');
    }

    const passwordHash = await this.passwordHasher.hash(dto.password);

    const user = new User({
      email: dto.email,
      passwordHash,
      fullName: dto.fullName,
      phone: dto.phone,
      role: UserRole.MEMBER,
      skills: dto.skills ?? [],
      vulnerabilities: dto.vulnerabilities ?? [],
      resources: dto.resources ?? [],
      latitude: dto.latitude,
      longitude: dto.longitude,
      isVerified: false,
    });

    return this.userRepository.save(user);
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/RegisterUser.cs
using Hestia.Application.DTOs;
using Hestia.Application.Interfaces;
using Hestia.Domain.Entities;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class RegisterUser
{
    private readonly IUserRepository _userRepository;
    private readonly IPasswordHasher _passwordHasher;

    public RegisterUser(IUserRepository userRepository, IPasswordHasher passwordHasher)
    {
        _userRepository = userRepository;
        _passwordHasher = passwordHasher;
    }

    public async Task<User> ExecuteAsync(RegisterRequest dto, CancellationToken ct = default)
    {
        var existing = await _userRepository.FindByEmailAsync(dto.Email, ct);
        if (existing is not null)
            throw new ApplicationException("User with this email already exists");

        var hash = _passwordHasher.Hash(dto.Password);

        var user = new User(
            email: dto.Email,
            passwordHash: hash,
            fullName: dto.FullName,
            phone: dto.Phone,
            skills: dto.Skills,
            vulnerabilities: dto.Vulnerabilities,
            resources: dto.Resources,
            latitude: dto.Latitude,
            longitude: dto.Longitude);

        return await _userRepository.SaveAsync(user, ct);
    }
}
